Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -203,35 +203,26 @@ private void PullerLoop(CancellationToken cancellationToken)
 
                 using (new StopwatchDisposable(o => this.Validator.PerformanceCounter.AddBlockFetchingTime(o)))
                 {
+                    // Save the current consensus tip to later check if it changed.
+                    ChainedBlock consensusTip = this.Tip;
+
                     // This method will block until the next block is downloaded.
                     blockValidationContext.Block = this.Puller.NextBlock(cancellationToken);
 
                     if (blockValidationContext.Block == null)
                     {
                         lock (this.consensusLock)
                         {
-                            this.logger.LogTrace("No block received from puller due to reorganization, rewinding.");
-                            ChainedBlock lastTip = this.Tip;
-
-                            CancellationToken token = this.nodeLifetime.ApplicationStopping;
+                            this.logger.LogTrace("No block received from puller due to reorganization.");
 
-                            ChainedBlock rewinded = null;
-                            while (rewinded == null)
+                            if (!consensusTip.Equals(this.Tip))
```

Comment:
This condition is to avoid rewinding in case a block was mined in the mean while?

If we assume that `consensusTip` will always be in sync with puller.Tip then this is fine, in case the chain advanced the reorg will just be triggered in the next block.

Reply:
Yes, to avoid rewinding in case a block was delivered from our miner. I.e. I guess puller could return null as for reorg, but the reason was not that was intended in the puller - peer changed the chain, but that our miner added a block.

So I'm just saving the state of consensus before i go to puller, then i check if that changed or not. If not, I know that the reorg is from outside. If yes, then I know that reorg is from inside.